### PR TITLE
devicestate: support seeding from a base snap instead of core

### PR DESCRIPTION
**Pull request by mvo5**
This PR makes seeding work for models that only use a base snap for booting and have no "core" snap installed.

This is based on https://github.com/snapcore/snapd/pull/5274

The relevant commit is 0610b2e

### CODE REVIEW

File: overlord/devicestate/firstboot.go [hunk 1]
```diff
@@ -92,15 +91,13 @@ func populateStateFromSeedImpl(st *state.State) ([]*state.TaskSet, error) {
...
-		return trivialSeeding(st, markSeeded), nil
-	}
 	if err != nil {
 		return nil, err
 	}
 
+	assertSeedDir := filepath.Join(dirs.SnapSeedDir, "assertions")
 	seedYamlFile := filepath.Join(dirs.SnapSeedDir, "seed.yaml")
-	if release.OnClassic && !osutil.FileExists(seedYamlFile) {
+	if release.OnClassic && (!osutil.FileExists(seedYamlFile) || !osutil.FileExists(assertSeedDir)) {
```
Comment (pedronis) [inline overlord/devicestate/firstboot.go]: having seed.yaml without an explicit model file is intentionally not supported

Comment (pedronis) [reply overlord/devicestate/firstboot.go]: mmh, this code preserve that but I find the new order of checks not easy to parse

**Comment by mvo5** [comment]
@pedronis Yeah, I reverted this particular change. The previous incarnation of this code was using the model early but this is no longer needed thanks to the agreement to hang the config onto "core".

File: overlord/devicestate/firstboot.go [hunk 2]
```diff
@@ -128,18 +129,25 @@ func populateStateFromSeedImpl(st *state.State) ([]*state.TaskSet, error) {
 	tsAll := []*state.TaskSet{}
 	configTss := []*state.TaskSet{}
 
-	// if there are snaps to seed, core needs to be seeded too
+	baseSnap := "core"
+	if model.Base() != "" {
+		baseSnap = model.Base()
+	}
+	// if there are snaps to seed, core/base needs to be seeded too
 	if len(seed.Snaps) != 0 {
```
Comment (pedronis) [inline overlord/devicestate/firstboot.go]: to be consistent if base is not "core" we need code to make sure "snapd" is installed first here, so   snapd, then base etc

File: overlord/devicestate/firstboot.go [hunk 3]
```diff
@@ -128,52 +129,58 @@ func populateStateFromSeedImpl(st *state.State) ([]*state.TaskSet, error) {
...
-	if len(seed.Snaps) != 0 {
-		coreSeed := seeding["core"]
-		if coreSeed == nil {
-			return nil, fmt.Errorf("cannot proceed without seeding core")
+	baseSnap := "core"
+	if model.Base() != "" {
+		baseSnap = model.Base()
+	}
+
+	installSeedHelper := func(snapName string, last int) error {
```
Comment (pedronis) [inline overlord/devicestate/firstboot.go]: can we call this maybe  "installSeedEssential" ?  "Helper" doesn't say much

File: overlord/snapstate/handlers.go [hunk 4]
```diff
@@ -187,9 +187,14 @@ func (m *SnapManager) doPrerequisites(t *state.Task, _ *tomb.Tomb) error {
 		return err
 	}
 
-	// core/ubuntu-core can not have prerequisites
-	snapName := snapsup.Name()
-	if snapName == defaultCoreSnapName || snapName == "ubuntu-core" {
+	// os/base/kernel/gadget cannot have prerequisites other
+	// than the models default base (or core) which is installed anyway
+	switch snapsup.Type {
+	case snap.TypeOS, snap.TypeBase, snap.TypeKernel, snap.TypeGadget:
```
Comment (pedronis) [inline overlord/snapstate/handlers.go]: do we have a test for this?

Comment (mvo5) [reply overlord/snapstate/handlers.go]: Yes, I will add one.

**Pull request merged**